gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/InputStreamDataWriter.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.copy.writer;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.data.management.copy.FileAwareInputStream;
+import gobblin.data.management.util.PathUtils;
+import gobblin.util.ForkOperatorUtils;
+import gobblin.util.HadoopUtils;
+import gobblin.util.WriterUtils;
+import gobblin.writer.DataWriter;
+
+import java.io.IOException;
+import java.net.URI;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.io.Closer;
+
+/**
+ * A {@link DataWriter} to write {@link FileAwareInputStream}
+ */
+@Slf4j
+public class InputStreamDataWriter implements DataWriter<FileAwareInputStream> {
+
+  protected long bytesWritten = 0;
+  protected long filesWritten = 0;
+  protected final State state;
+  protected final FileSystem fs;
+  protected final Path stagingDir;
+  protected final Path outputDir;
+  protected Closer closer = Closer.create();
+
+  public InputStreamDataWriter(State state, int numBranches, int branchId) throws IOException {
+    this.state = state;
+
+    Configuration conf = new Configuration();
+    String uri =
+        this.state
+            .getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, numBranches,
+                branchId), ConfigurationKeys.LOCAL_FS_URI);
+
+    this.fs = FileSystem.get(URI.create(uri), conf);
+    this.stagingDir = WriterUtils.getWriterStagingDir(state, numBranches, branchId);
+    this.outputDir =
+        new Path(state.getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_OUTPUT_DIR,
+            numBranches, branchId)));
+  }
+
+  @Override
+  public void write(FileAwareInputStream fileAwareInputStream) throws IOException {
+    closer.register(fileAwareInputStream.getInputStream());
+    Path stagingFile = getStagingFilePath(fileAwareInputStream.getFile());
+    this.fs.mkdirs(stagingFile.getParent());
+
+    FSDataOutputStream os = closer.register(fs.create(stagingFile, true));
+    IOUtils.copyLarge(fileAwareInputStream.getInputStream(), os);
+    os.close();
+    filesWritten++;
+
+    setFilePermissions(fileAwareInputStream.getFile());
+  }
+
+  /**
+   * Sets the owner/group and permission for the file in the task staging directory
+   *
+   * @param file
+   * @throws IOException
+   */
+  protected void setFilePermissions(CopyableFile file) throws IOException {
```

Comment:
We should also set the metadata of the parent directories if available (under ancestorsOwnerAndPermission)